jaxrs-service/src/main/java/com/quorum/tessera/p2p/PartyInfoResource.java

```diff
@@ -1,55 +1,143 @@
+        this.enclave = requireNonNull(enclave);
+        this.payloadEncoder = payloadEncoder;
+    }
+
+    public PartyInfoResource(
+        final PartyInfoService partyInfoService,
+        final PartyInfoParser partyInfoParser,
+        Client restClient,
+        Enclave enclave) {
+
+        this(partyInfoService, partyInfoParser, restClient, enclave, PayloadEncoder.create());
     }
 
     /**
      * Allows node information to be retrieved in a specific encoded form
      * including other node URLS and public key to URL mappings
      *
      * @param payload The encoded node information from the requester
-     * @return the merged node information from this node, which may contain new information
+     * @return the merged node information from this node, which may contain new
+     * information
      */
     @POST
     @Consumes(MediaType.APPLICATION_OCTET_STREAM)
     @Produces(MediaType.APPLICATION_OCTET_STREAM)
     @ApiOperation(value = "Request public key/url of other nodes", produces = "public keylist/url")
-    @ApiResponses({@ApiResponse(code = 200, message = "Encoded PartyInfo Data", response = byte[].class)})
+    @ApiResponses({
+        @ApiResponse(code = 200, message = "Encoded PartyInfo Data", response = byte[].class)})
     public Response partyInfo(@ApiParam(required = true) final byte[] payload) {
 
         final PartyInfo partyInfo = partyInfoParser.from(payload);
 
-        final PartyInfo updatedPartyInfo = partyInfoService.updatePartyInfo(partyInfo);
+        //Start validation stuff
+        PublicKey sender = enclave.defaultPublicKey();
+
+        final String url = partyInfo.getUrl();
+
+        final String dataToEncrypt = UUID.randomUUID().toString();
+
+        Predicate<Recipient> isValidRecipientKey = r -> {
+
+            PublicKey key = r.getKey();
+            final EncodedPayload encodedPayload = enclave.encryptPayload(dataToEncrypt.getBytes(), sender, Arrays.asList(key));
+
+            byte[] encodedPayloadData = payloadEncoder.encode(encodedPayload);
+
+            Response response = restClient.target(url)
+                .path("partyinfo")
+                .path("validate")
+                .request()
+                .post(Entity.entity(encodedPayloadData, MediaType.APPLICATION_OCTET_STREAM));
+            
+            String unencodedValidationData = response.readEntity(String.class);
+            
+            boolean isValid = Objects.equals(unencodedValidationData, dataToEncrypt);
+            if(!isValid) {
+                LOGGER.warn("Invalid key found {} recipient will be ignored.",url);
+            }
+            
+            return isValid;
+
+        };
+
+        Predicate<Recipient> isSendingUrl = r -> r.getUrl().equalsIgnoreCase(url);
+        
+        //Validate caller and treat no valid certs as security issue.  
+        partyInfo.getRecipients()
+            .stream()
+            .filter(isValidRecipientKey.and(isSendingUrl))
```

Comment:
We should check if the URL matches first before trying to make a call out
```suggestion
            .filter(isSendingUrl.and(isValidRecipientKey))
```